pkg/client/client.go

```diff
@@ -352,6 +385,146 @@ func (c *serviceImpl) doRequest(method, path string, reqpl, respl any) error {
 	return nil
 }
 
+const openConnections = 10
+
+type filesUploader struct {
+	Client         *http.Client
+	EndpointURL    string
+	Filenames      []string
+	ChunkSizeBytes int64
+	DumpOut        io.Writer
+}
+
+type fileInfo struct {
+	Name        string
+	TotalChunks int
+}
+
+func (u *filesUploader) Upload() error {
+	var totalChunks int
+	var infos []fileInfo
+	for _, name := range u.Filenames {
+		stat, err := os.Stat(name)
+		if err != nil {
+			return nil
+		}
+		info := fileInfo{
+			Name:        name,
+			TotalChunks: int(math.Ceil(float64(stat.Size()) / float64(u.ChunkSizeBytes))),
+		}
+		infos = append(infos, info)
+		totalChunks += info.TotalChunks
+	}
+	jobsChan := make(chan *uploadChunkJob, openConnections)
+	resultsChan := make(chan error, 1)
+	defer close(jobsChan)
+	defer close(resultsChan)
+	for i := 0; i < openConnections; i++ {
+		w := &uploadChunkWorker{
+			Client:      u.Client,
+			EndpointURL: u.EndpointURL,
+			DumpOut:     u.DumpOut,
+			JobsChan:    jobsChan,
+			ResultsChan: resultsChan,
+		}
+		go w.Start()
+	}
+	go func() {
+		for _, info := range infos {
+			for i := 0; i < info.TotalChunks; i++ {
+				j := &uploadChunkJob{
+					Filename:       info.Name,
+					ChunkNumber:    i + 1,
+					TotalChunks:    info.TotalChunks,
+					ChunkSizeBytes: u.ChunkSizeBytes,
+				}
+				jobsChan <- j
+			}
+		}
+	}()
+	for i := 0; i < totalChunks; i++ {
+		if err := <-resultsChan; err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+type uploadChunkJob struct {
+	Filename string
+	// Chunk numbers are 1-indexed.
+	ChunkNumber    int
+	TotalChunks    int
+	ChunkSizeBytes int64
+}
+
+type uploadChunkWorker struct {
+	Client      *http.Client
+	EndpointURL string
+	DumpOut     io.Writer
+	JobsChan    <-chan *uploadChunkJob
+	ResultsChan chan<- error
+
+	bodyBuff bytes.Buffer
+}
+
+func (w *uploadChunkWorker) Start() {
+	for j := range w.JobsChan {
+		w.ResultsChan <- w.upload(j)
+	}
+}
+
+func (w *uploadChunkWorker) upload(job *uploadChunkJob) error {
+	file, err := os.Open(job.Filename)
+	if err != nil {
+		return err
+	}
+	defer file.Close()
+	if _, err := file.Seek(int64(job.ChunkNumber-1)*job.ChunkSizeBytes, 0); err != nil {
+		return err
+	}
+	mpWriter := multipart.NewWriter(&w.bodyBuff)
+	name := fmt.Sprintf("%s_%d_%d.chunked", filepath.Base(job.Filename), job.ChunkNumber, job.TotalChunks)
+	part, err := mpWriter.CreateFormFile("file", name)
+	if err != nil {
+		return err
+	}
+	if job.ChunkNumber < job.TotalChunks {
+		if _, err = io.CopyN(part, file, job.ChunkSizeBytes); err != nil {
+			return err
+		}
+	} else {
+		if _, err = io.Copy(part, file); err != nil {
+			return err
+		}
+	}
+	mpWriter.Close()
+	// client trace to log whether the request's underlying tcp connection was re-used
+	clientTrace := &httptrace.ClientTrace{
```

Comment:
is this a leftover from debugging? If that dump line is intentional it should include more information, like chunk number out chunk count and file name. That way we'll get a better idea of how many connections were not reused.

Reply:
added more info as suggested.